src/event.rs

```diff
@@ -0,0 +1,610 @@
+use lightning::util::persist::KVStorePersister;
+use lightning::util::ser::{MaybeReadable, Readable, ReadableArgs, Writeable, Writer};
+
+use bitcoin::secp256k1::Secp256k1;
+use rand::{thread_rng, Rng};
+use std::collections::{hash_map, VecDeque};
+use std::sync::{Arc, Condvar, Mutex};
+use std::thread;
+use std::time::Duration;
+
+/// The event queue will be persisted under this key.
+pub(crate) const EVENTS_PERSISTENCE_KEY: &str = "events";
+
+/// An event emitted by [`LdkLite`] that should be handled by the user.
+///
+/// This wraps an event of a specific type and will more often than not wrap a LDK event accessible
+/// via the `inner` field.
+///
+/// [`LdkLite`]: [`crate::LdkLite`]
+#[derive(Debug, Clone)]
+pub enum LdkLiteEvent {
+	/// A payment we sent was successful.
+	PaymentSuccessful(PaymentSuccessfulEvent),
+	/// A payment we sent has failed.
+	PaymentFailed(PaymentFailedEvent),
+	/// A payment has been received.
+	PaymentReceived(PaymentReceivedEvent),
+	// TODO: Implement after a corresponding LDK event is added.
+	//ChannelOpened(ChannelOpenedEvent),
+	/// A channel has been closed.
+	ChannelClosed(ChannelClosedEvent),
+	// TODO: Implement on-chain events when better integrating with BDK wallet sync.
+	//OnChainPaymentSent(OnChainPaymentSentEvent),
+	//OnChainPaymentReceived(OnChainPaymentReceivedEvent),
+}
+
+trait EventType {
+	const TYPE: u8;
+}
+
+fn write_event<T: EventType + Writeable, W: Writer>(
+	event: &T, writer: &mut W,
+) -> Result<(), lightning::io::Error> {
+	T::TYPE.write(writer)?;
+	event.write(writer)?;
+	Ok(())
+}
+
+impl Readable for LdkLiteEvent {
+	fn read<R: lightning::io::Read>(
+		reader: &mut R,
+	) -> Result<Self, lightning::ln::msgs::DecodeError> {
+		match Readable::read(reader)? {
+			PaymentSuccessfulEvent::TYPE => Ok(Self::PaymentSuccessful(Readable::read(reader)?)),
+			PaymentFailedEvent::TYPE => Ok(Self::PaymentFailed(Readable::read(reader)?)),
+			PaymentReceivedEvent::TYPE => Ok(Self::PaymentReceived(Readable::read(reader)?)),
+			// ChannelOpenedEvent::TYPE => {
+			// TODO ChannelOpened
+			//}
+			ChannelClosedEvent::TYPE => Ok(Self::ChannelClosed(Readable::read(reader)?)),
+			//5u8 => {
+			// TODO OnChainPaymentSent
+			//}
+			//6u8 => {
+			// TODO OnChainPaymentReceived
+			//}
+			_ => Err(lightning::ln::msgs::DecodeError::InvalidValue),
+		}
+	}
+}
+
+impl Writeable for LdkLiteEvent {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), lightning::io::Error> {
+		match self {
+			Self::PaymentSuccessful(event) => write_event(event, writer)?,
+			Self::PaymentFailed(event) => write_event(event, writer)?,
+			Self::PaymentReceived(event) => write_event(event, writer)?,
+			Self::ChannelClosed(event) => write_event(event, writer)?,
+		}
+		Ok(())
+	}
+}
+
+/// A payment we sent was successful.
+#[derive(Debug, Clone)]
+pub struct PaymentSuccessfulEvent {
+	/// The hash of the payment.
+	pub payment_hash: PaymentHash,
+	/// The wrapped LDK event.
+	pub inner: ldk_events::Event,
```

Comment:
Don't have a strong opinion yet, but we could make `inner` private and add accessors exposing whatever details are relevant.

Reply:
So, the thinking was to expose the bare minimum of fields (and possibly additional data) via the simplified LDKLite events, but allow a user to 'drill deeper' by providing the wrapped event. So the main API should def. be the first-level fields, but if that is not enough for some reason, we provide a fallback option (rather than start going down the rabbit hole of making more and more data first-level fields and hence increasing the API surface again). This could also somewhat provide a migration path from LDKLite -> 'full' LDK.

Reply:
Now dropped wrapping upstream events with https://github.com/lightningdevkit/ldk-lite/pull/10/commits/0cd90b5d38e59cee3571dc361e62d555b6497563.